edge/pkg/edged/volume/csi/csi_attacher.go

```diff
@@ -0,0 +1,447 @@
+			klog.Error(log("Failed to check CSIDriver for %s: %s", driverName, err))
+		} else {
+			if skip {
+				// This volume is not attachable, pretend it's attached
+				attached[spec] = true
+				continue
+			}
+		}
+
+		attachID := getAttachmentName(volumeHandle, driverName, string(nodeName))
+		klog.V(4).Info(log("probing attachment status for VolumeAttachment %v", attachID))
+		attach, err := c.k8s.StorageV1().VolumeAttachments().Get(attachID, meta.GetOptions{})
+		if err != nil {
+			attached[spec] = false
+			klog.Error(log("attacher.VolumesAreAttached failed for attach.ID=%v: %v", attachID, err))
+			continue
+		}
+		klog.V(4).Info(log("attacher.VolumesAreAttached attachment [%v] has status.attached=%t", attachID, attach.Status.Attached))
+		attached[spec] = attach.Status.Attached
+	}
+
+	return attached, nil
+}
+
+func (c *csiAttacher) GetDeviceMountPath(spec *volume.Spec) (string, error) {
+	klog.V(4).Info(log("attacher.GetDeviceMountPath(%v)", spec))
+	deviceMountPath, err := makeDeviceMountPath(c.plugin, spec)
+	if err != nil {
+		klog.Error(log("attacher.GetDeviceMountPath failed to make device mount path: %v", err))
+		return "", err
+	}
+	klog.V(4).Infof("attacher.GetDeviceMountPath succeeded, deviceMountPath: %s", deviceMountPath)
+	return deviceMountPath, nil
+}
+
+func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMountPath string) (err error) {
+	klog.V(4).Infof(log("attacher.MountDevice(%s, %s)", devicePath, deviceMountPath))
+
+	if deviceMountPath == "" {
+		err = fmt.Errorf("attacher.MountDevice failed, deviceMountPath is empty")
+		return err
+	}
+
+	mounted, err := isDirMounted(c.plugin, deviceMountPath)
+	if err != nil {
+		klog.Error(log("attacher.MountDevice failed while checking mount status for dir [%s]", deviceMountPath))
+		return err
+	}
+
+	if mounted {
+		klog.V(4).Info(log("attacher.MountDevice skipping mount, dir already mounted [%s]", deviceMountPath))
+		return nil
+	}
+
+	// Setup
+	if spec == nil {
+		return fmt.Errorf("attacher.MountDevice failed, spec is nil")
+	}
+	csiSource, err := getPVSourceFromSpec(spec)
+	if err != nil {
+		klog.Error(log("attacher.MountDevice failed to get CSIPersistentVolumeSource: %v", err))
+		return err
+	}
+
+	// Store volume metadata for UnmountDevice. Keep it around even if the
+	// driver does not support NodeStage, UnmountDevice still needs it.
+	if err = os.MkdirAll(deviceMountPath, 0750); err != nil {
+		klog.Error(log("attacher.MountDevice failed to create dir %#v:  %v", deviceMountPath, err))
+		return err
+	}
+	klog.V(4).Info(log("created target path successfully [%s]", deviceMountPath))
+	dataDir := filepath.Dir(deviceMountPath)
+	data := map[string]string{
+		volDataKey.volHandle:  csiSource.VolumeHandle,
+		volDataKey.driverName: csiSource.Driver,
+	}
+	if err = saveVolumeData(dataDir, volDataFileName, data); err != nil {
+		klog.Error(log("failed to save volume info data: %v", err))
+		if cleanerr := os.RemoveAll(dataDir); err != nil {
```

Comment:
can this be moved to the defer block as well ?

Reply:
Let's keep it as the native kubernetes csi did.
This is handled by the specified error for `saveVolumeData`.